### PR TITLE
feat: implement `pure/List.mo`

### LINKED ISSUE #102: Implementation: `pure/List` data structure

### PR DESCRIPTION
**Pull request by ggreif**
(Partially) resolves #102.

TODO
- [ ] unit tests, esp. checking for stack exhaustion
- [ ] finalise interface for `contains`

### CODE REVIEW

File: src/pure/List.mo [hunk 1]
```diff
@@ -11,147 +20,778 @@ module {
...
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func partition<T>(list : List<T>, f : T -> Bool) : (List<T>, List<T>) =
+    switch list {
+      case null (null, null);
+      case (?(h, t)) {
+        let left = f h;
+        let (l, r) = partition(t, f);
```
Comment (luc-blaeser) [inline src/pure/List.mo:281]: Could be a problem with stack overflow for longer list. We had such issues in the past. I suggest using a loop.

File: src/pure/List.mo [hunk 2]
```diff
@@ -11,147 +20,778 @@ module {
...
+  ///
+  /// Runtime: O(n)
+  ///
+  /// Space: O(n)
+  public func split<T>(list : List<T>, n : Nat) : (List<T>, List<T>) =
+    if (n == 0) (null, list)
+    else switch list {
+      case null (null, null);
+      case (?(h, t)) {
+        let (l1, l2) = split(t, n - 1);
```
Comment (luc-blaeser) [inline src/pure/List.mo:671]: Could be an issue with stack overflow for long lists.

**Review by luc-blaeser — COMMENTED** [review]
Nice work.
I suggest adding unit tests, especially also a larger-scaled case with a 10,000+ list to check that no stack overflow happens for longer lists.

**Review by luc-blaeser — COMMENTED** [review]
Approved. Thanks. Some follow up aspects

**Review by rvanasa — APPROVED** [review]
